webapp/handlers.py

```diff
@@ -102,14 +102,20 @@ def add_headers(response):
             if flask.session:
                 response.headers["Cache-Control"] = "private"
             else:
+                cache_control = {"public"}
+
+                if "/static/images/badges" in flask.request.url:
+                    cache_control.update({"max-age=31556952"})
+                else:
+                    cache_control.update(
```

Comment:
Yes we could do this, but it's just slightly bit risky.

We don't actually generate all static URLs with hashes, what we do instead is allow you to request in the template to generate a URL with a query string that is a hash of the file's content, so:
```
<link rel="stylesheet" href="{{ static_url('css/styles.css') }}" />
```

Becomes:

```
<link rel="stylesheet" href="/static/css/styles.css?v=dcf8316" />
```

This structure means that for `styles.css` itself, an infinite cache shouldn't matter as the hash will change when the file changes.

However, it's still perfectly possible to link to assets directly from templates (e.g. `href="/static/css/styles.css"`), in which case an infinite cache would be a significant problem.

The idea was that, for any file that might changes (so, basically all CSS and JavaScript files) we should use the `static_url` function - I don't know if we do use it everywhere. Maybe this could be improved such that it's impossible to request static files directly, to avoid the risk of people not using the `static_url` function where they should? We could explore that.

However, that's all a bit tangential to this particular issue. The intention was actually that the local `/static/` directory was always intended *only* for local, regularly changing files that are tighly coupled to the project, like local CSS and JavaScript. All other "static" files should be served from the assets server assets.ubuntu.com.

assets.ubuntu.com already includes hashes baked into the URL and has year-long caches, and it can do this without risk because it explicitly prohibits replacing the contents of an asset at the same URL.

In this case, where we're serving static images that are the same for everybody (https://snapcraft.io/static/images/badges/en/snap-store-black.svg and https://snapcraft.io/static/images/badges/en/snap-store-white.svg), these should simply be served from the assets server.

Reply:
@nottrobin Initially we served those images from asset server, but this leads to some bad URL 'usability'.

Asset server URLs contain hash (which is great for cashing), but makes URLs less user friendly.
Lets say we have a badge at: https://assets.ubuntu.com/v1/ab1e8e6c-EN-snap-store-black-uneditable.svg

First of all it's a snapcraft badge served from ubuntu.com domain. It's not a huge issue for our regular assets, but these badges' URLs are visible in HTML/Markdown snippets we ask users to use, so we want them to look nice.

Then this file URL looks like it can be changed. Can I change `EN` to `FR` to get other language? No - because I don't know the hash. Same for changing `black`/`white`.

And also we'd like to be able to serve those badges from one consistent URL regardless of their 'version'. So if we update the translation, we don't need to update the URLs (and user snippets).

Reply:
You can create assets on the assets server with URLs that you specify, e.g. https://assets.ubuntu.com/v1/vanilla-framework-version-0.0.15.css. It's not possible through the web manager currently, but I can show anyone how to do it. This would both make the URL more user friendly and solve your "en/fr" problem.

The domain problem... well we could change assets.ubuntu.com to assets.canonical.com - and all non-OS-specific domains and products started being called "Canonical" rather than "Ubuntu". If there was a will for us to do that we certainly could. We could even just alias it for now, so the same assets are available at both assets.ubuntu.com and asset.canonical.com. In my opinion, that should be good enough for Snapcraft peeps - I believe association with Canonical is a boon to the Snapcraft brand.

Reply:
So I just did a quick test:
```
/static/images/badges/de/snap-store-black.svg
  Local transferred:    11.23kb
  Local total size:     28.56kb
  Redirect transferred: 14.51kb
  Redirect total size:  14.11kb
```
Not sure what to read from it really. Doesn't seem too drastic to redirect, but could add quite an overhead to the code complexity on top of the small size bump.
We could just use assets.ubuntu.com urls, but the initial reason for not using it is to have prettier URLs that people can copy and paste.
Would squid negate this conversation?

Reply:
No I can't see how Squid would help. I did mention a few ways above in which assets.ubuntu.com URLs could be improved. Some right away, some with work (e.g. assets.canonical.com). Maybe with those improvements you would no longer feel the need to redirect?